### PR TITLE
Plane: fix terrain following in loiter and QRTL

### LINKED ISSUE #14668: Plane: Loiter does not use TERRAIN Follow
While adding a terrain follow disable feature for CRUISE/FBWB/LOITER modes, while keeping it active for AUTO/GUIDED/RTL modes, discovered in SITL that LOITER does not use terrain following....I had setup a 1km WP  radius, flying  over mountains SE of Canberra and was surprised to find that LOITER never terrain follows...either WIKI needs to be changed, or code does...
if a code change occurs, please assure that the Non Auto Mode Terrain Following Disable switch I am implementing could disable it, as it does for CRUISE/FBWB in its PR #14666 

**Platform**
[  ] All
[  ] AntennaTracker
[  ] Copter
[ X ] Plane
[  ] Rover
[  ] Submarine

### PR DESCRIPTION
**Pull request by IamPete1**
This fixes and adds a test for loiter terrain following in plane. I think that is fairly clear.

Slightly less clear is the fix for QRTL. It seems there is a split between some-places using `plane.next_WP_loc.terrain_alt` as the flag for terrain following and others using `target_altitude.terrain_following`. Its all very confusing, I would be un-surprised if terrain following is also broken in a bunch of other modes.

Fixes #14668

### CODE REVIEW

**Comment by IamPete1** [comment]
The new auto test is failing to load terrain correctly, some tests have comments saying they load MAVProxy to make it more reliable, that does not seem to help locally tho. @peterbarker any ideas?

File: ArduPlane/quadplane.cpp [hunk 1]
```diff
@@ -2617,6 +2617,12 @@ void QuadPlane::vtol_position_controller(void)
                                                      plane.auto_state.wp_proportion,
                                                      0, 1);
             }
+#if AP_TERRAIN_AVAILABLE
+            float terrain_altitude_offset;
+            if (plane.next_WP_loc.terrain_alt && plane.terrain.height_terrain_difference_home(terrain_altitude_offset, true)) {
```
Comment (tridge) [inline ArduPlane/quadplane.cpp:2622]: how does this handle rally points?

**Review by tridge — CHANGES_REQUESTED** [review]

File: Tools/autotest/arduplane.py [hunk 2]
```diff
@@ -2180,6 +2180,36 @@ def fly_terrain_mission(self):
 
         self.fly_mission("ap-terrain.txt", mission_timeout=600)
 
+    def test_loiter_terrain(self):
+        default_rad = self.get_parameter("WP_LOITER_RAD")
```
Comment (peterbarker) [inline Tools/autotest/arduplane.py:2184]: Shouldn't we really start MAVProxy here to supply the terrain?

**Comment by tridge** [comment]
we could have the pre-generated terrain dat file in git, and copy to terrain directory in sitl

**Comment by IamPete1** [comment]
I think I finally have this right now. Plane was previously calling stuff relative to terrain but in reality it was still relative to home. Its all a massive mess that it impossible to understand. I would be completely un-surprised if terrain following was broken in a bunch of other modes.

File: ArduPlane/quadplane.cpp [hunk 3]
```diff
@@ -2623,16 +2626,29 @@ void QuadPlane::vtol_position_controller(void)
         }
         if (plane.control_mode == &plane.mode_qrtl || plane.control_mode == &plane.mode_guided || vtol_loiter_auto) {
             plane.ahrs.get_position(plane.current_loc);
-            float target_altitude = plane.next_WP_loc.alt;
-            if (poscontrol.slow_descent) {
+            int32_t target_altitude_cm;
+            if (!plane.next_WP_loc.get_alt_cm(Location::AltFrame::ABOVE_HOME,target_altitude_cm)) {
+                break;
+            }
+            if (poscontrol.slow_descent && 0) {
```
Comment (tridge) [inline ArduPlane/quadplane.cpp:2633]: did you really want that && 0 ??

**Review by tridge — CHANGES_REQUESTED** [review]
the patch looks half finished?

File: ArduPlane/quadplane.cpp (hunk 3 above)
Comment (IamPete1) [reply ArduPlane/quadplane.cpp:2633]: no, left over from testing, removed.

File: ArduPlane/quadplane.cpp [hunk 4]
```diff
@@ -2626,16 +2629,29 @@ void QuadPlane::vtol_position_controller(void)
...
+            }
+#if AP_TERRAIN_AVAILABLE
+            float terrain_altitude_offset_cm;
+            if (plane.next_WP_loc.terrain_alt && plane.terrain.height_terrain_difference_home(terrain_altitude_offset_cm, true)) {
+                // Climb if current terrain is above home, target_altitude_cm is reltive to home
+                target_altitude_cm += MAX(terrain_altitude_offset_cm*100,0);
             }
-            adjust_alt_target(target_altitude - plane.home.alt);
+#endif
+            pos_control->set_alt_target_with_slew(target_altitude_cm, plane.G_Dt);
```
Comment (tridge) [inline ArduPlane/quadplane.cpp:2654]: might need diff between home and origin ?

File: ArduPlane/quadplane.cpp [hunk 5]
```diff
@@ -2819,11 +2835,18 @@ void QuadPlane::init_qrtl(void)
...
     poscontrol.speed_scale = 0;
     pos_control->set_desired_accel_xy(0.0f, 0.0f);
     pos_control->init_xy_controller();
+    int32_t from_alt;
+    int32_t to_alt;
+    if (plane.current_loc.get_alt_cm(Location::AltFrame::ABSOLUTE,from_alt) && plane.next_WP_loc.get_alt_cm(Location::AltFrame::ABSOLUTE,to_alt)) {
+        poscontrol.slow_descent = from_alt > to_alt;
+        return;
+    }
+    // defualt back to old method
```
Comment (tridge) [inline ArduPlane/quadplane.cpp:2848]: speeeling

File: ArduPlane/quadplane.cpp [hunk 6]
```diff
@@ -2626,16 +2629,29 @@ void QuadPlane::vtol_position_controller(void)
...
+                int32_t prev_alt;
+                if (plane.prev_WP_loc.get_alt_cm(Location::AltFrame::ABOVE_HOME,prev_alt)) {
+                    target_altitude_cm = linear_interpolate(prev_alt,
+                                                         target_altitude_cm,
+                                                         plane.auto_state.wp_proportion,
+                                                         0, 1);
+                }
+            }
+#if AP_TERRAIN_AVAILABLE
+            float terrain_altitude_offset_cm;
```
Comment (tridge) [inline ArduPlane/quadplane.cpp:2648]: should be _m not _cm

**Review by tridge — CHANGES_REQUESTED** [review]

**Review by tridge — APPROVED** [review]

**Pull request merged**